pkg/agent/plugin/workloadattestor/systemd/systemd_posix_test.go

```diff
@@ -0,0 +1,146 @@
+//go:build !windows
+// +build !windows
+
+package systemd
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+	"github.com/sirupsen/logrus/hooks/test"
+	"github.com/spiffe/spire/pkg/agent/plugin/workloadattestor"
+	"github.com/spiffe/spire/test/plugintest"
+	"github.com/spiffe/spire/test/spiretest"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc/codes"
+)
+
+var (
+	ctx = context.Background()
+)
+
+func TestPlugin(t *testing.T) {
+	spiretest.Run(t, new(Suite))
+}
+
+type Suite struct {
+	spiretest.Suite
+
+	log     logrus.FieldLogger
+	logHook *test.Hook
+}
+
+func (s *Suite) SetupTest() {
+	log, logHook := test.NewNullLogger()
+	s.log = log
+	s.logHook = logHook
+}
+
+func (s *Suite) TestAttest() {
+	testCases := []struct {
+		name           string
+		pid            int
+		selectorValues []string
+		expectCode     codes.Code
+		expectMsg      string
+		expectLogs     []spiretest.LogEntry
+	}{
+		{
+			name:       "get unit id",
+			pid:        1,
+			expectCode: codes.OK,
+		},
+		{
+			name:       "fail to get fragment path",
+			pid:        1,
+			expectCode: codes.OK,
+		},
+		{
+			name:       "fail to get unit id",
+			pid:        2,
+			expectCode: codes.Internal,
+			expectMsg:  "workloadattestor(systemd): unable to get unit Id for process",
+		},
+		{
+			name:       "fail to get fragment path",
+			pid:        2,
+			expectCode: codes.Internal,
+			expectMsg:  "workloadattestor(systemd): unable to get unit FragmentPath for process",
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+		s.T().Run(testCase.name, func(t *testing.T) {
+			defer s.logHook.Reset()
+
+			p := s.loadPlugin(t)
+			selectors, err := p.Attest(ctx, testCase.pid)
+			spiretest.RequireGRPCStatus(t, err, testCase.expectCode, testCase.expectMsg)
+			if testCase.expectCode != codes.OK {
+				require.Nil(t, selectors)
+				return
+			}
+
+			require.NoError(t, err)
+			require.NotNil(t, selectors)
+			var selectorValues []string
+			for _, selector := range selectors {
+				require.Equal(t, "systemd", selector.Type)
+				selectorValues = append(selectorValues, selector.Value)
+			}
+
+			require.Equal(t, testCase.selectorValues, selectorValues)
+			spiretest.AssertLogs(t, s.logHook.AllEntries(), testCase.expectLogs)
+		})
+	}
+}
+
+func (s *Suite) loadPlugin(t *testing.T) workloadattestor.WorkloadAttestor {
+	p := s.newPlugin()
+
+	v1 := new(workloadattestor.V1)
+	plugintest.Load(t, builtin(p), v1,
+		plugintest.Log(s.log))
+	return v1
+}
+
+func (s *Suite) newPlugin() *Plugin {
+	p := New()
+	p.getUnitInfo = func(ctx context.Context, pid uint) (unitInfo, error) {
+		return newFakeUnit(pid), nil
+	}
+	return p
+}
+
+type fakeUnit struct {
+	pid int32
+}
+
+func (u fakeUnit) Id() (string, error) {
```

Comment:
I believe this code is not compiling for two reasons:
- On line 127 and 129, we are returning `nil` for a string field
- `p` on line 123 is not defined here nor in outer scope, you should correct the receiver here to `u`